README.md
# Basic information

## Online documentation

https://rclUE.readthedocs.io/en/devel/

## Supported versions

- Ubuntu 20.04
- Unreal Engine 4.27
- ROS2 Foxy
- Clang: 11.0.1 (from UE)

Please setup UE4 by following [Linux Quick Start](https://docs.unrealengine.com/4.27/en-US/SharingAndReleasing/Linux/BeginnerLinuxDeveloper/SettingUpAnUnrealWorkflow/)

## Branches

### master

This build of the plugin is based on ROS2 Foxy and has been tested on Ubuntu 20.

This branch currently represents the release branch.

### Other branches

Experimental and WIP

# RclUE and ROS2

## Description
- We use ros2 'foxy' lightweighted (not all binaries are included). Source/ThirdParty/ros folder is fully autogenerated by Scripts/update_ros2.py (check script to see what is actually includes).
- ros includes [UE_msgs](https://github.com/rapyuta-robotics/UE_msgs)
- Ros2 middleware, Product name: eProsima Fast DDS, License: Apache 2, RMW implementation: rmw_fastrtps_cpp
- to use this plugin change the world settings to use meters as units instead of the default centimeters
- within the Unreal Editor: Edit->Plugins, search and enable for `rclc`

## Windows is currently unsupported

# Getting Started

The plugin folder contains a video "Example_BP_PubSub.mp4" demonstrating how to setup a PubSub example in Blueprint.

An example setup using this plugin can be found at [turtlebot3-UE](https://github.com/rapyuta-robotics/turtlebot3-UE)

# Notes on working with ROS2 and UE4

- rcl and void\* types cannot be managed by UE4 (no UPROPERTY) and therefore can't be used directly in Blueprint. Whenever access to these variables is needed, the user should write a class to wrap it and all of their handling must be done in C++.
- some basic numerical types are not natively supported in Blueprint (e.g. double, unsigned int). In order to use these, a workaround is needed (a plugin implementing those types for BP, a modified UE4 or a custom implementation).
- In autogenerated messages, the method MsgToString() should be implemented by the user as its current purpose is to help debugging.

# How to update ROS inside RclUE
Currently there is a Scripts/update_ros2.py for automatic build and update ROS2 (foxy) libraries. To update ROS:
- Install 'ros2' (foxy) (until section 'Build the code in the workspace') [Building ROS 2 on Ubuntu Linux](https://docs.ros.org/en/foxy/Installation/Ubuntu-Development-Setup.html)
- Reinstall python package due to [issue](https://github.com/ros-visualization/qt_gui_core/issues/212):
```
sudo apt remove shiboken2 libshiboken2-dev libshiboken2-py3-5.14
pip3 install --user shiboken2
```
- Apply patch for ros2 to avoid setting LD_LIBRARY_PATH environment variable to the dynamic libs paths:
```
src/ros2/rcpputils/src/find_library.cpp:
in function find_library_path():
---//return "";
+++return "lib" + library_name + ".so";
```
- Clone ['rclc' package](https://github.com/ros2/rclc/tree/foxy), branch 'foxy' (it's not a part of official ros2)
- Clone ['ue_msgs'](https://github.com/rapyuta-robotics/UE_msgs) (Rapyuta-Robotics project) 
- Run Scripts/update_ros2.py script with arguments:
```
python3 update_ros2.py <Path to UE4> <Path to ROS2 foxy> <Path to rclc package> <Path to UE_msgs> <Path to project, which includes RclUE plugin>
For example:
python3 update_ros2.py /home/vilkun/UE/UnrealEngine /home/vilkun/ros2_foxy /home/vilkun/work/build_foxy/rclc /home/vilkun/work/build_foxy/UE_msgs /home/vilkun/work/turtlebot3-UE
```
Notes:
- Script also updates Rapyuta-Robotics UE_msgs libraries inside RclUE.
- Please check the scripts for more details about how it deals with libraries versions in their SONAME (UE4 can't deal with it, see 2 potential fix ways)

# Documentation

## Tools

documentation is built with three tools

- [doxygen](http://www.doxygen.org)
- [sphinx](http://www.sphinx-doc.org)
- [breathe](https://breathe.readthedocs.io)

## Locally build

1. install tools in #tools section.
2. build
   ```
   cd docs
   make --always-make html
   ```
3. Open following in your browser.
   - Sphinx at `file:///<path to cloned repo>/docs/source/_build/html/index.html`
   - Original doxygen output at `file:///<path to cloned repo>/docs/source/_build/html/doxygen_generated/html/index.html`

# Maintainer

yu.okamoto@rapyuta-robotics.com


CHANGELOG.md
# Changelog for rclUE repository
## 0.0.6 ##
* ROS2 Utils: Add publish template
* rclUE.Build.cs Copy ros shared libs to TargetOutputDir to co-locate with other non-rclUE ue-built ros libs

## 0.0.5 ##
* Update ros libs and all msg/srv/action sources generated from latest Ros foxy & UE_tools #55
* Move scripts building ros libs for UE to UE_tools
## 0.0.4 ##
* Add ROS2 Utils: String UE <-> ROS conversion, topic subscription/publication

## 0.0.3 ##
* Add ros2-foxy libraries automatic build and update  

## 0.0.2 ##
* Add sphinx documentation
* Add SpawnEntities service

## 0.0.1 ##
* Add CHANGELOG
* Update maintainers


Scripts/update_ros.py
#!/usr/bin/env python3
'''
How to update ros:

- Install 'ros2' (foxy) (until section 'Build the code in the workspace') https://docs.ros.org/en/foxy/Installation/Ubuntu-Development-Setup.html
- Reinstall python package due to issues https://github.com/ros-visualization/qt_gui_core/issues/212:
sudo apt remove shiboken2 libshiboken2-dev libshiboken2-py3-5.14
pip3 install --user shiboken2
- Make replacement for ros2 to avoid LD_LIBRARY_PATH dependency issue:
src/ros2/rcpputils/src/find_library.cpp:
in function find_library_path():
---//return "";
+++return "lib" + library_name + ".so";
- Clone 'rclc' package (which is not a part of official ros2), branch 'foxy' https://github.com/ros2/rclc/tree/foxy
- Clone 'ue_msgs' (RR project) https://github.com/rapyuta-robotics/UE_msgs
- Run this script with providing main path arguments, for example:
python3 update_ros.py /home/vilkun/UE/UnrealEngine /home/vilkun/ros2_foxy /home/vilkun/work/build_foxy/rclc /home/vilkun/work/build_foxy/UE_msgs /home/vilkun/work/turtlebot3-UE

Notes:
- 304 packages finished, 1 'rviz_ogre_vendor' error, 6 packages not processed at the end of the ros-foxy build - is okay. Also 1 'Failed to find ROS1', and several 'connext' errors, and 'rviz' errors during 'local_setup.bash' - are okay too.
- we rename SONAME of libs with version by 'patchelf' since UE4.27 can't deal with that in LinuxToolChain.cs, see more in RenameLibsWithVersion().
- script uses project path (which includes this plugin) just to invalidate binaries, and forces IDE to clean cache and link them again.
'''

import os, sys, shutil, re, time, subprocess
from libs_utils import *

# Main arguments:
UE = sys.argv[1]
ros = sys.argv[2]
rclc = sys.argv[3]
ue_msgs = sys.argv[4]
projectPath = sys.argv[5]
# 'rcl' will include 'rclc'
allowed_spaces = [ \
    'action_msgs', \
    'builtin_interfaces', \
    'example_interfaces', \
    'fastcdr', \
    'fastrtps', \
    'geometry_msgs', \
    'nav_msgs', \
    'rcl', \
    'rcutils', \
    'rcpputils', \
    'rmw', \
    'rosgraph_msgs', \
    'rosidl', \
    'sensor_msgs', \
    'std_msgs', \
    'tf2_msgs', \
    'tracetools', \
    'unique_identifier_msgs' ]
rr_spaces = ['ue_msgs']
allowed_spaces.extend(rr_spaces)
not_allowed_spaces = [ \
    '.so.', \
    'python', \
    'rclcpp', 
    'rclpy', 
    'rcl_logging_log4cxx', \
    'rcl_logging_noop', \
    'rclc_examples', \
    'rclc_parameter', \
    'connext', \
    'cyclonedds', \
    'action_tutorials', \
    'actionlib_msgs', \
    'composition_interfaces', \
    'diagnostic_msgs', \
    'lifecycle_msgs', \
    'logging_demo', \
    'map_msgs', \
    'move_base_msgs', \
    'pendulum_msgs', \
    'shape_msgs', \
    'statistics_msgs', \
    'std_srvs', \
    'stereo_msgs', \
    'test_msgs', \
    'trajectory_msgs', \
    'visualization_msgs', \
    'turtlesim', \
    'rmw_fastrtps_dynamic_cpp', \
    'rosidl_cmake', \
    'rosidl_default', \
    'rosidl_generator_cpp', \
    'rosidl_generator_dds_idl', \
    'rosidl_generator_py', \
    'rosidl_runtime_cpp', \
    'rosidl_runtime_py', \
    'sensor_msgs_py', \
    'tf2_sensor_msgs', \
    'tracetools_', \
    'Codec_', \
    'Plugin_', \
    'RenderSystem_', \
    '_test_type_support']

rclUE =  projectPath + '/Plugins/rclUE'
buildRosScript = rclUE + '/Scripts/build_ros_libs.sh'
rclUEBinaries = rclUE + '/Binaries'
rclUERos = rclUE + '/Source/ThirdParty/ros'
rclUERosInclude = rclUERos + '/include'
rclUERosLib = rclUERos + '/lib'
rclUEBuildCS =  rclUE + '/Source/rclUE/rclUE.Build.cs'
projectPathBinaries = projectPath + '/Binaries' 
infoRosOutput = rclUE + '/Scripts/info_ros'

# autogenerated folders
renamedSuffix = '_renamed'
rosBuild  = ros + '/build'
rosInstall  = ros + '/install'
rclcBuild = rclc + '/build'
rclcInstall = rclc + '/install'
ue_msgsBuild = ue_msgs + '/build'
ue_msgsInstall = ue_msgs + '/install'

if __name__ == '__main__':
    start = time.time()
    
    print('Building ros...')
    os.system('chmod +x ' + buildRosScript)
    os.system('bash ' + buildRosScript + ' ' + UE + ' ' + ros + ' ' + rclc + ' ' + ue_msgs)

    # to be sure that we don't use libraries from this paths:
    rosBuild = RenameDir(rosBuild, renamedSuffix)
    rosInstall = RenameDir(rosInstall, renamedSuffix)
    rclcBuild = RenameDir(rclcBuild, renamedSuffix)
    rclcInstall = RenameDir(rclcInstall, renamedSuffix)
    ue_msgsBuild = RenameDir(ue_msgsBuild, renamedSuffix)
    ue_msgsInstall = RenameDir(ue_msgsInstall, renamedSuffix)

    # See more comments inside function description
    if os.path.isdir(rclUERos):
        shutil.rmtree(rclUERos)
        
    os.makedirs(rclUERosInclude)
    os.makedirs(rclUERosLib)

    print('Grabbing includes...')
    GrabIncludes(rosInstall, rclUERosInclude, allowed_spaces)
    GrabIncludes(rclcInstall, rclUERosInclude, allowed_spaces)
    GrabIncludes(ue_msgsInstall, rclUERosInclude, allowed_spaces)
    CleanIncludes(rclUERosInclude, not_allowed_spaces)

    print('Applying includes patch...')
    os.system('cd .. ; git apply Scripts/ros.patch')

    print('Grabbing libs...')
    GrabLibs(rosBuild, rclUERosLib, allowed_spaces)
    GrabLibs(rclcBuild, rclUERosLib, allowed_spaces)
    GrabLibs(ue_msgsBuild, rclUERosLib, allowed_spaces)
    CleanLibs(rclUERosLib, not_allowed_spaces)

    RenameLibsWithVersion(rclUERosLib)
    SetRPATH(rclUERosLib)
    InvalidateBinaries(projectPathBinaries, rclUEBinaries, rclUEBuildCS)

    # You also can try this:
    # CreateInfoForAll('objdump -x', rclUERosLib, infoRosOutput) # see also 'ldd'
    # CheckLibs(rclUERosLib)
    print('Done. Time:', '{:.2f}'.format(time.time() - start), '(sec)')